xds/src/main/java/io/grpc/xds/XdsClusterResource.java

```diff
@@ -282,16 +241,41 @@ private static StructOrError<CdsUpdate.Builder> parseNonAggregateCluster(
           + ": transport-socket-matches not supported.");
     }
     if (cluster.hasTransportSocket()) {
-      if (!TRANSPORT_SOCKET_NAME_TLS.equals(cluster.getTransportSocket().getName())) {
-        return StructOrError.fromError("transport-socket with name "
-            + cluster.getTransportSocket().getName() + " not supported.");
+      String transportSocketName = cluster.getTransportSocket().getName();
+
+      if (!TRANSPORT_SOCKET_NAME_TLS.equals(transportSocketName) &&
+          (isEnabledXdsHttpConnect && !TRANSPORT_SOCKET_NAME_HTTP11_PROXY.equals(transportSocketName))) {
+        return StructOrError.fromError(
+            "Transport socket with name " + transportSocketName + " not supported.");
+      }
+
+      Any tlsConfig;
+      if (isEnabledXdsHttpConnect && TRANSPORT_SOCKET_NAME_HTTP11_PROXY.equals(transportSocketName)) {
+        isHttp11ProxyEnabled = true;
+        try {
+          tlsConfig = unpackCompatibleType(
+              cluster.getTransportSocket().getTypedConfig(),
+              io.envoyproxy.envoy.extensions.transport_sockets.http_11_proxy.v3
+                  .Http11ProxyUpstreamTransport.class,
+              TRANSPORT_SOCKET_NAME_HTTP11_PROXY, null
+          ).getTransportSocket().getTypedConfig();
+        } catch (InvalidProtocolBufferException e) {
+          return StructOrError.fromError(
+              "Cluster " + clusterName + ": malformed Http11ProxyUpstreamTransport: " + e);
+        } catch (ClassCastException e) {
+          return StructOrError.fromError(
+              "Cluster " + clusterName + ": invalid transport_socket type in Http11ProxyUpstreamTransport");
+        }
+      } else {
+        tlsConfig = cluster.getTransportSocket().getTypedConfig();
       }
+
       try {
         upstreamTlsContext = UpstreamTlsContext.fromEnvoyProtoUpstreamTlsContext(
             validateUpstreamTlsContext(
-                unpackCompatibleType(cluster.getTransportSocket().getTypedConfig(),
-                io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext.class,
-                TYPE_URL_UPSTREAM_TLS_CONTEXT, TYPE_URL_UPSTREAM_TLS_CONTEXT_V2),
+                unpackCompatibleType(tlsConfig,
```

Comment:
You aren't checking that this is a TRANSPORT_SOCKET_NAME_TLS.

Reply:
You've protected this block, but you don't fail if transport socket inside Http11ProxyUpstreamTransport is something other than not-present and TRANSPORT_SOCKET_NAME_TLS. I suggest you split the HTTP1.1 out to be a bit separate and restore the old TLS flow.

```
boolean hasTransportSocket = cluster.hasTransportSocket();
TransportSocket transportSocket = cluster.getTransportSocket();
if (hasTransportSocket && isEnabledXdsHttpConnect && TRANSPORT_SOCKET_NAME_HTTP11_PROXY.equals(transportSocket.getName())) {
  // etc
  hasTransportSocket = wrappedTransportSocket.hasTransportSocket();
  transportSocket = wrappedTransportSocket.getTransportSocket();
}
if (hasTransportSocket) {
  // the old TLS code
  ...
}
```